include/graphblas/reference/blas3.hpp

```diff
@@ -916,6 +916,206 @@ namespace grb {
+				return PANIC;
+			} else {
+				return FAILED;
+			}
+		}
+
+		grb::Monoid<
+			grb::operators::left_assign< OutputType >,
+			grb::identities::zero
+		> monoid;
+
+		RC ret = SUCCESS;
+		if( phase == EXECUTE ) {
+			ret = grb::clear( A );
+		}
+		assert( nnz( A ) == 0 );
+
+		auto &CRS_raw = internal::getCRS( A );
+		auto &CCS_raw = internal::getCCS( A );
+
+		const InputType1 * __restrict__ const x = internal::getRaw( u );
+		const InputType2 * __restrict__ const y = internal::getRaw( v );
+		char * arr = nullptr;
+		char * buf = nullptr;
+		OutputType * valbuf = nullptr;
+		internal::getMatrixBuffers( arr, buf, valbuf, 1, A );
+		config::NonzeroIndexType * A_col_index = internal::template
+			getReferenceBuffer< typename config::NonzeroIndexType >( ncols + 1 );
+
+		
+		internal::Coordinates< reference > coors;
+		coors.set( arr, false, buf, ncols );
+
+		CRS_raw.col_start[ 0 ] = 0;
+		for( size_t j = 0; j <= ncols; ++j ) {
+			CCS_raw.col_start[ j ] = 0;
+		}
+
+		
+		nzc = 0;
+		for( size_t i = 0; i < nrows; ++i ) {
+			if( internal::getCoordinates( u ).assigned( i ) ) {
+				for( auto k = mask_raw.col_start[ i ]; k < mask_raw.col_start[ i + 1 ]; ++k ) {
+					const size_t k_col = mask_raw.row_index[ k ];
+					if( internal::getCoordinates( v ).assigned( k_col ) ) {
+						(void) ++nzc;
+						(void) ++CCS_raw.col_start[ k_col + 1 ];
+					}
+				}
+			}
+			CRS_raw.col_start[ i + 1 ] = nzc;
+		}
+
+
+		A_col_index[ 0 ] = 0;
+		for( size_t j = 1; j < ncols; ++j ) {
+			CCS_raw.col_start[ j + 1 ] += CCS_raw.col_start[ j ];
+			A_col_index[ j ] = 0;
+		}
+
+
+		const size_t old_nzc = nzc;
+
+		// use previously computed CCS offset array to update CCS during the
+		// computational phase
+		nzc = 0;
+		for( size_t i = 0; i < nrows; ++i ) {
+			coors.clear();
+			if( internal::getCoordinates( u ).assigned( i ) ) {
+				for( auto k = mask_raw.col_start[ i ]; k < mask_raw.col_start[ i + 1 ]; ++k ) {
+					const size_t k_col = mask_raw.row_index[ k ];
+					if( internal::getCoordinates( v ).assigned( k_col ) ) {
+						coors.assign( k_col );
+						valbuf[ k_col ] = monoid.template getIdentity< OutputType >();
+						(void) grb::apply( valbuf[ k_col ],
+							x[i],
+							y[k_col],
+							mul );
+					}
+				}
+			}
+			for( size_t k = 0; k < coors.nonzeroes(); ++k ) {
+				assert( nzc < old_nzc );
+				const size_t j = coors.index( k );
+				// update CRS
+				CRS_raw.row_index[ nzc ] = j;
+				CRS_raw.setValue( nzc, valbuf[ j ] );
+				// update CCS
+				const size_t CCS_index = A_col_index[ j ]++ + CCS_raw.col_start[ j ];
+				CCS_raw.row_index[ CCS_index ] = i;
+				CCS_raw.setValue( CCS_index, valbuf[ j ] );
+				// update count
+				(void) ++nzc;
+			}
+			CRS_raw.col_start[ i + 1 ] = nzc;
+		}
+
+		for( size_t j = 0; j < ncols; ++j ) {
+			assert( CCS_raw.col_start[ j + 1 ] - CCS_raw.col_start[ j ] ==
+				A_col_index[ j ] );
+		}
+		assert( nzc == old_nzc );
+		
+
+		internal::setCurrentNonzeroes( A, nzc );
+		
+		return ret;
+	}
```

Comment:
Comments on the whole function:
- Add openMP support
- Wrap any CCS logic into `if( !( descr & descriptors::force_row_major ) )`

Reply:
Done, thanks!